### PR TITLE
E net

### PR DESCRIPTION
**Pull request by CharlesAuthier**

### CODE REVIEW

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+It is divided into several stages, as highlighted by horizontal lines in the table and the first digit after each block name.
+
+They adopt a view of ResNet that describes them as having a single main branch and extensions with convolutional filters that separate from it, and then merge back with an element-wise addition.  Each block consists of three convolutional layers: a 1x1 projection that reduces the dimensionality, a main convolutional layer (conv in figure bellow), and a 1x1 expansion.
+A Batch Normalization  and PReLU are also used between all convolutions.
+Just as in the original paper of the ResNet, they refer to these as bottleneck modules.
+If the bottleneck is downsampling, a max pooling layer is added to the main branch.
+
+Also, the first 1x1 projection is replaced with a 2x2 convolution with stride 2 in both dimensions.
+`convis` either a regular, dilated or full convolution (also known as deconvolution or fractionally strided convolution) with 3x3 filters.
+For the regularizer, thay use Spatial Dropout, with p= 0.01 before bottleneck 2.0, and p= 0.1 afterwards.
```
Comment (jhlegarreta) [inline deep-learning/_posts/2019-05-06-ENet.md]: thay: they

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+The architecture of this network is in the table bellow.
+It is divided into several stages, as highlighted by horizontal lines in the table and the first digit after each block name.
+
+They adopt a view of ResNet that describes them as having a single main branch and extensions with convolutional filters that separate from it, and then merge back with an element-wise addition.  Each block consists of three convolutional layers: a 1x1 projection that reduces the dimensionality, a main convolutional layer (conv in figure bellow), and a 1x1 expansion.
+A Batch Normalization  and PReLU are also used between all convolutions.
+Just as in the original paper of the ResNet, they refer to these as bottleneck modules.
+If the bottleneck is downsampling, a max pooling layer is added to the main branch.
+
+Also, the first 1x1 projection is replaced with a 2x2 convolution with stride 2 in both dimensions.
+`convis` either a regular, dilated or full convolution (also known as deconvolution or fractionally strided convolution) with 3x3 filters.
```
Comment (jhlegarreta) [inline deep-learning/_posts/2019-05-06-ENet.md]: The sentence is unclear: does the sentence describe what `convis` is?

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+For the regularizer, thay use Spatial Dropout, with p= 0.01 before bottleneck 2.0, and p= 0.1 afterwards.
+The initial stage contains a single block, Stage 1 consists of 5 bottleneck blocks, while stage 2 and 3 have the same structure, with the exception that stage 3 does not downsample the input at the beginning (they omit the 0th bottleneck).
+These three first stages are the encoder.
```
Comment (jhlegarreta) [inline deep-learning/_posts/2019-05-06-ENet.md]: May be mention this before describing its parts?

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+---
+layout: review
+title:  "ENet: A Deep Neural Network Architecture for Real-Time Semantic Segmentation"
+tags:   deep-learning, segmentation, benchmarking, CNN
```
Comment (jhlegarreta) [inline deep-learning/_posts/2019-05-06-ENet.md]: Tags need not to (must not?) be comma-separated, whitespace separation is enough, right?

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+author: Charles Authier
+pdf:  https://arxiv.org/pdf/1606.02147.pdf
+cite:
+  authors: "Adam Paszke, Abhishek Chaurasia, Sangpil Kim, Eugenio Culurciello"
+  title:   "ENet: A Deep Neural Network Architecture for Real-Time Semantic Segmentation"
+  venue:   "arXiv preprint arXiv:1606.02147 2016/6/7"
+---
+
+## Idea
+Recent deep neural networks aimed to the real-time pixel-wise  semantic  segmentation task have the disadvantage of requiring a large number of floating point operations and have long run-times that hinder their usability.
```
Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: "aimed at real-time"

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+In this paper, they propose a new deep neural network architecture named **ENet** for *efficient neural network*, created specifically for tasks requiring low latency operation.
+They claim that the ENet is up to 18×faster, requires 75×less FLOPs, has 79×less parameters, and provides similar or better accuracy to existing models like the **SegNet**.
+Thay have tested it on CamVid, Cityscapes and SUN datasets.
```
Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: They

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+![](/deep-learning/images/ENet/enet.png)
+
+## Network
+The architecture of this network is in the table bellow.
```
Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: below

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+It is divided into several stages, as highlighted by horizontal lines in the table and the first digit after each block name.
+
+They adopt a view of ResNet that describes them as having a single main branch and extensions with convolutional filters that separate from it, and then merge back with an element-wise addition.  Each block consists of three convolutional layers: a 1x1 projection that reduces the dimensionality, a main convolutional layer (conv in figure bellow), and a 1x1 expansion.
```
Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: below

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+Also, the first 1x1 projection is replaced with a 2x2 convolution with stride 2 in both dimensions.
+`conv` (convolution) is either a regular, dilated or full convolution (also known as deconvolution or fractionally strided convolution) with 3x3 filters.
+For the regularizer, they use Spatial Dropout, with p= 0.01 before bottleneck 2.0, and p= 0.1 afterwards.
+The initial stage contains a single block, Stage 1 consists of 5 bottleneck blocks, while stage 2 and 3 have the same structure, with the exception that stage 3 does not downsample the input at the beginning (they omit the 0th bottleneck).
+
+**They did not use bias terms in any of the projections, in order to reduce the number of kernel calls and overall memory operations**, as cuDNN uses separate kernels for convolution and bias addition.
+This choice didn’t have any impact on the accuracy.
+Between each convolutional layer and following non-linearity they use Batch Normalization.
+In the decoder maxpooling is replaced with max unpooling, and padding is replaced with spatial convolution without bias.
```
Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: ~In~ The decoder

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+![](/deep-learning/images/ENet/network.png)
+
+![](/deep-learning/images/ENet/networklayers.png)
+
+In section 4 of the article, they justified every decision about every part of the network, for more information see this section.
+
+## Results
+# Inference Time
+The table bellow compares inference time for a single input frame of varying resolution. They also report the number of frames per second that can be processed.
```
Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: below

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+Dashes indicate that they could not obtain a measurement, due to lack of memory.
+ENet is significantly faster than SegNet, providing high frame rates for real-time applications and allowing for practical use of very deep neural network models with encoder-decoder architecture.
+
+![](/deep-learning/images/ENet/inferencetime.png)
+
+# Hardware Requirements
+The table bellow reports a comparison of number of floating point operations and parameters used by different models.
```
Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: below

Comment (AntoineTheb) [inline deep-learning/_posts/2019-05-06-ENet.md]: reports a comparison ~of~ ~number~ of floating point operations

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
+---
+layout: review
+title:  "ENet: A Deep Neural Network Architecture for Real-Time Semantic Segmentation"
+tags:   deep-learning, segmentation, benchmarking, CNN
+author: Charles Authier
+pdf:  https://arxiv.org/pdf/1606.02147.pdf
+cite:
+  authors: "Adam Paszke, Abhishek Chaurasia, Sangpil Kim, Eugenio Culurciello"
+  title:   "ENet: A Deep Neural Network Architecture for Real-Time Semantic Segmentation"
+  venue:   "arXiv preprint arXiv:1606.02147 2016/6/7"
```
Comment (jodoin) [inline deep-learning/_posts/2019-05-06-ENet.md]: put real ICLR reference

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+---
+
+## Idea
+Recent deep neural networks aimed to the real-time pixel-wise  semantic  segmentation task have the disadvantage of requiring a large number of floating point operations and have long run-times that hinder their usability.
+In this paper, they propose a new deep neural network architecture named **ENet** for *efficient neural network*, created specifically for tasks requiring low latency operation.
```
Comment (jodoin) [inline deep-learning/_posts/2019-05-06-ENet.md]: they authors propose

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+![](/deep-learning/images/ENet/enet.png)
+
+## Network
+The architecture of this network is in the table bellow.
+It is divided into several stages, as highlighted by horizontal lines in the table and the first digit after each block name.
+
+They adopt a view of ResNet that describes them as having a single main branch and extensions with convolutional filters that separate from it, and then merge back with an element-wise addition.  Each block consists of three convolutional layers: a 1x1 projection that reduces the dimensionality, a main convolutional layer (conv in figure bellow), and a 1x1 expansion.
+A Batch Normalization  and PReLU are also used between all convolutions.
+Just as in the original paper of the ResNet, they refer to these as bottleneck modules.
```
Comment (jodoin) [inline deep-learning/_posts/2019-05-06-ENet.md]: original  ResNet paper,

File: deep-learning/_posts/2019-05-06-ENet.md (hunk 4 above)
Comment (jhlegarreta) [reply deep-learning/_posts/2019-05-06-ENet.md]: Done in a037b83.

File: deep-learning/_posts/2019-05-06-ENet.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+cite:
+  authors: "Adam Paszke, Abhishek Chaurasia, Sangpil Kim, Eugenio Culurciello"
+  title:   "ENet: A Deep Neural Network Architecture for Real-Time Semantic Segmentation"
+  venue:   "arXiv preprint arXiv:1606.02147 2016/6/7"
+---
+
+## Idea
+Recent deep neural networks aimed to the real-time pixel-wise  semantic  segmentation task have the disadvantage of requiring a large number of floating point operations and have long run-times that hinder their usability.
+In this paper, they propose a new deep neural network architecture named **ENet** for *efficient neural network*, created specifically for tasks requiring low latency operation.
+They claim that the ENet is up to 18×faster, requires 75×less FLOPs, has 79×less parameters, and provides similar orbetter accuracy to existing models like the **SegNet**.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply deep-learning/_posts/2019-05-06-ENet.md]: Done in a037b83.

File: deep-learning/_posts/2019-05-06-ENet.md (hunk 1 above)
Comment (jhlegarreta) [reply deep-learning/_posts/2019-05-06-ENet.md]: Done in a037b83.

File: deep-learning/_posts/2019-05-06-ENet.md (hunk 3 above)
Comment (jhlegarreta) [reply deep-learning/_posts/2019-05-06-ENet.md]: Addressed in a037b83.

File: deep-learning/_posts/2019-05-06-ENet.md (hunk 2 above)
Comment (jhlegarreta) [reply deep-learning/_posts/2019-05-06-ENet.md]: Addressed in a037b83.

**Pull request merged**